### PR TITLE
[Enhancement] array_expr return constant column

### PR DESCRIPTION
**Pull request by Seaven**
Why I'm doing:

What I'm doing:
optimize array_expr use large memory

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exprs/array_expr.cpp [hunk 1]
```diff
@@ -56,15 +60,19 @@ class ArrayExpr final : public Expr {
...
             }
             curr_offset += num_elements;
             array_offsets->append(curr_offset);
         }
 
-        return ArrayColumn::create(std::move(array_elements), std::move(array_offsets));
+        auto ptr = ArrayColumn::create(std::move(array_elements), std::move(array_offsets));
+        if (all_const) {
+            return ConstColumn::create(std::move(ptr), output_rows);
+        }
```
Comment (fzhedu) [inline be/src/exprs/array_expr.cpp:74]: would it result only null?

File: be/src/exprs/cast_nested.cpp [hunk 2]
```diff
@@ -120,6 +121,9 @@ StatusOr<ColumnPtr> CastArrayExpr::evaluate_checked(ExprContext* context, Chunk*
             ArrayColumn::create(std::move(casted_element_column),
                                 ColumnHelper::as_column<UInt32Column>(array_column->offsets_column()->clone_shared()));
     RETURN_IF_ERROR(casted_array->unfold_const_children(_type));
+    if (orig_column->is_constant()) {
```
Comment (fzhedu) [inline be/src/exprs/cast_nested.cpp:124]: would it be only null?

File: be/src/exprs/math_functions.cpp [hunk 3]
```diff
@@ -768,6 +768,17 @@ StatusOr<ColumnPtr> MathFunctions::cosine_similarity(FunctionContext* context, c
...
+    if (base->is_constant()) {
+        auto* const_column = down_cast<const ConstColumn*>(base);
+        const_column->data_column()->assign(base->size(), 0);
+        base = const_column->data_column().get();
+    }
+    if (target->is_constant()) {
+        auto* const_column = down_cast<const ConstColumn*>(target);
+        const_column->data_column()->assign(target->size(), 0);
+        target = const_column->data_column().get();
+    }
```
Comment (fzhedu) [inline be/src/exprs/math_functions.cpp:780]: use unfold_const() function?

File: be/src/exprs/array_expr.cpp (hunk 1 above)
Comment (Seaven) [reply be/src/exprs/array_expr.cpp:74]: no, must not only_null

File: be/src/exprs/cast_nested.cpp (hunk 2 above)
Comment (Seaven) [reply be/src/exprs/cast_nested.cpp:124]: may be, but was handle it in line 103

File: be/src/exprs/math_functions.cpp (hunk 3 above)
Comment (Seaven) [reply be/src/exprs/math_functions.cpp:780]: here must not be `only_null`, so I only check `is_constant`

**Review by fzhedu — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by trueeyu** [comment]
@Mergifyio backport branch-3.2

**Comment by trueeyu** [comment]
https://github.com/Mergifyio backport branch-3.1

**Pull request merged**